In this project you will be creating a presentation that will be delivered to a government board on behalf of concerned citizens.
You are working on behalf of community groups who believe there should be more oversight of prescription opioids.
You will want each slide to be simple yet informative.  You can use any charts you choose and you can assume the board understands
simple statistics and correlation coefficients.  You have 10 minutes to present and will be expected to answer basic questions.
At the end of the presentation you should recommend some areas that could benefit from further data collection and analysis.

--------------------------------------------------------------------------------------------------------------------------------------------

Before you plan the presentation, start by answering some of these questions.  They will help you get an idea of some possible directions you can go.  Not all these answers need to be in the presentation and other analysis can be included.

NOTE:  Not all of these questions may be answerable.  Sometimes requests will be made for answers the data can not provide.
Practice learning why types of questions can, and can not, be answered with the data as-is.

In [1]:
from sqlalchemy import create_engine, text
!pip install psycopg2-binary
import pandas as pd

sqlalchemy to fetch prescribers database

In [2]:
database_name = 'prescribers'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"
engine = create_engine(connection_string)

test a query

In [3]:
query = 'SELECT * FROM population'

In [4]:
with engine.connect() as connection:
    population = pd.read_sql(query,  con = connection)
population

,fipscounty,population
0,47017,28137.0
1,47023,17097.0
2,47039,11681.0
3,47037,678322.0
4,47087,11573.0
...,...,...
90,47089,52887.0
91,47141,75565.0
92,47153,14654.0
93,47155,95523.0


--------------------------------------------------------------------------------------------------------------------------------------------

1. Deaths over time.

In [5]:
query = 'SELECT * FROM overdose_deaths'

with engine.connect() as connection:
    overdose_deaths= pd.read_sql(query, con = connection)

overdose_deaths

,overdose_deaths,year,fipscounty
0,135,2015,47157
1,150,2016,47157
2,159,2017,47157
3,123,2018,47157
4,122,2015,47093
...,...,...,...
375,0,2018,47017
376,1,2015,47007
377,2,2016,47007
378,2,2017,47007


a. How has total overdose deaths changed over time?

In [6]:
overdose_deaths.groupby("year")["overdose_deaths"].sum().reset_index()

,year,overdose_deaths
0,2015,1033
1,2016,1186
2,2017,1267
3,2018,1304


b. How have overdose deaths changed over time for Davidson and Shelby counties.

In [7]:
davidson_shelby = """SELECT * 
FROM overdose_deaths 
INNER JOIN fips_county ON overdose_deaths.fipscounty = CAST(fips_county.fipscounty AS int) 
WHERE county = 'DAVIDSON' OR county = 'SHELBY'"""

In [8]:
with engine.connect() as connection:
   davidson_shelby = pd.read_sql(davidson_shelby, con = connection)
davidson_shelby

,overdose_deaths,year,fipscounty,county,state,fipscounty,fipsstate
0,135,2015,47157,SHELBY,TN,47157,47
1,150,2016,47157,SHELBY,TN,47157,47
2,159,2017,47157,SHELBY,TN,47157,47
3,123,2018,47157,SHELBY,TN,47157,47
4,127,2015,47037,DAVIDSON,TN,47037,47
5,178,2016,47037,DAVIDSON,TN,47037,47
6,184,2017,47037,DAVIDSON,TN,47037,47
7,200,2018,47037,DAVIDSON,TN,47037,47


In [9]:
davidson_shelby.groupby("year")["overdose_deaths"].sum().reset_index()

,year,overdose_deaths
0,2015,262
1,2016,328
2,2017,343
3,2018,323


  c. Are there any counties in which overdose deaths are trending downward?

In [10]:
fips_overdose_query = """SELECT overdose_deaths, fipscounty, year
FROM overdose_deaths"""

with engine.connect() as connection:
    fips_overdose = pd.read_sql(fips_overdose_query, con=connection)
fips_overdose

,overdose_deaths,fipscounty,year
0,135,47157,2015
1,150,47157,2016
2,159,47157,2017
3,123,47157,2018
4,122,47093,2015
...,...,...,...
375,0,47017,2018
376,1,47007,2015
377,2,47007,2016
378,2,47007,2017


--------------------------------------------------------------------------------------------------------------------------------------------

2. Spending on opioids

a. What is the correlation between spending on opioids and overdose deaths?

In [11]:
from sqlalchemy import text

opioid_spending_query = (text("""WITH total_spent_npi AS (
  		SELECT ((total_drug_cost / total_day_supply) * 30 * total_30_day_fill_count) AS total_spent, npi
 	 	FROM prescription
  		LEFT JOIN drug ON prescription.drug_name = drug.drug_name
  		WHERE opioid_drug_flag = 'Y'
	)
SELECT 
  ROUND(SUM(total_spent_npi.total_spent * zip_fips.bus_ratio), 2) AS money_spent_opioid, 
  zip_fips.fipscounty
FROM total_spent_npi
INNER JOIN prescriber ON total_spent_npi.npi = prescriber.npi
INNER JOIN zip_fips ON prescriber.nppes_provider_zip5 = zip_fips.zip
WHERE fipscounty LIKE '47%'
GROUP BY zip_fips.fipscounty"""))

In [12]:
with engine.connect() as connection:
    opioid_spending_county = pd.read_sql(opioid_spending_query, con=connection)
opioid_spending_county

,money_spent_opioid,fipscounty
0,1801012.90,47001
1,269916.06,47003
2,697199.67,47005
3,98206.74,47007
4,1050004.81,47009
...,...,...
90,189407.80,47181
91,631100.96,47183
92,194737.46,47185
93,2217897.42,47187


The "zip_fips.bus_ratio" is the ratio of buisnesses in that zip code per county (ie accounts for overlap of zipcodes + counties)

fipscounty column is two different dtypes

In [13]:
opioid_spending_county['fipscounty'].dtype

dtype('O')

In [14]:
opioid_spending_county['fipscounty'].dtype

dtype('O')

--------------------------------------------------------------------------------------------------------------------------------------------

changing data types to str

In [15]:
fips_overdose['fipscounty'] = fips_overdose['fipscounty'].astype(str)
opioid_spending_county['fipscounty'] = opioid_spending_county['fipscounty'].astype(str)

--------------------------------------------------------------------------------------------------------------------------------------------

In [32]:
fips_overdose_index = fips_overdose.set_index('fipscounty')
spending_overdose = opioid_spending_county.join(fips_overdose_index, on='fipscounty', how='inner')
spending_overdose

,money_spent_opioid,fipscounty,overdose_deaths,year
0,1801012.90,47001,20,2015
0,1801012.90,47001,24,2016
0,1801012.90,47001,34,2017
0,1801012.90,47001,18,2018
1,269916.06,47003,8,2015
...,...,...,...,...
93,2217897.42,47187,30,2018
94,2078613.33,47189,26,2015
94,2078613.33,47189,27,2016
94,2078613.33,47189,26,2017


In [33]:
spending_overdose

,money_spent_opioid,fipscounty,overdose_deaths,year
0,1801012.90,47001,20,2015
0,1801012.90,47001,24,2016
0,1801012.90,47001,34,2017
0,1801012.90,47001,18,2018
1,269916.06,47003,8,2015
...,...,...,...,...
93,2217897.42,47187,30,2018
94,2078613.33,47189,26,2015
94,2078613.33,47189,27,2016
94,2078613.33,47189,26,2017


b. What is the ratio for spending on opioid vs non-opioid prescriptions?

spending on non-opioids (opioid_drug_flag = 'n')

In [18]:
from sqlalchemy import text

non_opioid_spending_query = text("""
WITH total_spent_npi AS (
    SELECT ((total_drug_cost / total_day_supply) * 30 * total_30_day_fill_count) AS total_spent, npi
    FROM prescription
    LEFT JOIN drug ON prescription.drug_name = drug.drug_name
    WHERE opioid_drug_flag = 'N'
)
SELECT 
    ROUND(SUM(total_spent_npi.total_spent * zip_fips.bus_ratio), 2) AS money_spent_non_opioid, 
    zip_fips.fipscounty
FROM total_spent_npi
INNER JOIN prescriber ON total_spent_npi.npi = prescriber.npi
INNER JOIN zip_fips ON prescriber.nppes_provider_zip5 = zip_fips.zip
WHERE fipscounty LIKE '47%'
GROUP BY zip_fips.fipscounty
""")

with engine.connect() as connection:
    non_opioid_spending_county = pd.read_sql(non_opioid_spending_query, con=connection)

non_opioid_spending_county

,money_spent_non_opioid,fipscounty
0,50020603.94,47001
1,10478185.98,47003
2,2847850.57,47005
3,2795144.45,47007
4,61534473.61,47009
...,...,...
90,5603432.48,47181
91,15227211.01,47183
92,6405561.79,47185
93,54655782.57,47187


In [ ]:
opioid_spending_county['fipscounty'] = opioid_spending_county['fipscounty'].astype(str)

In [ ]:
non_opioid_spending_county['fipscounty'] = non_opioid_spending_county['fipscounty'].astype(str)

merge does an inner join, only keeps countries present in both tables

In [ ]:
spending_ratio = opioid_spending_county.merge(non_opioid_spending_county, on='fipscounty')

compute for each country what fraction of prescription spening went to opioids. so opiod / total. total is opioid + non-opioid.

In [26]:
spending_ratio['opioid_ratio'] = spending_ratio['money_spent_opioid'] / (spending_ratio['money_spent_opioid'] + 
                                                                         spending_ratio['money_spent_non_opioid'])

spending_ratio

,money_spent_opioid,fipscounty,money_spent_non_opioid,opioid_ratio
0,1801012.90,47001,50020603.94,0.034754
1,269916.06,47003,10478185.98,0.025113
2,697199.67,47005,2847850.57,0.196668
3,98206.74,47007,2795144.45,0.033942
4,1050004.81,47009,61534473.61,0.016777
...,...,...,...,...
90,189407.80,47181,5603432.48,0.032697
91,631100.96,47183,15227211.01,0.039796
92,194737.46,47185,6405561.79,0.029504
93,2217897.42,47187,54655782.57,0.038997


c. Are those who spend a higher ratio on opioids suffering from more deaths?


In [38]:
# bring in overdose ratio
# calculate if there's a correlation between opioid_ratio and overdose_ratio
# This will only tell us correlation, not causation. So we can't fully answer this question because we don't know if
# higher opioid spending directly leads to more deaths.

result = pd.merge(spending_ratio, spending_overdose, on='fipscounty', how='inner')
result 

value = result['opioid_ratio'].corr(result['overdose_deaths'])
value

np.float64(-0.0993795642202032)

-0.09 corr value - insignificant correlation between spending and overdose deaths

--------------------------------------------------------------------------------------------------------------------------------------------

3. Per Capita

a. Which county has the highest overdose deaths per capita?

In [50]:
pop_result = pd.merge(population, result, on='fipscounty', how='inner')

pop_result["deaths_per_capita"] = (pop_result['overdose_deaths'])/(pop_result['population'])
pop_result

,fipscounty,population,money_spent_opioid_x,money_spent_non_opioid,opioid_ratio,money_spent_opioid_y,overdose_deaths,year,deaths_per_capita
0,47017,28137.0,741116.93,14727057.51,0.047912,741116.93,1,2015,0.000036
1,47017,28137.0,741116.93,14727057.51,0.047912,741116.93,1,2016,0.000036
2,47017,28137.0,741116.93,14727057.51,0.047912,741116.93,2,2017,0.000071
3,47017,28137.0,741116.93,14727057.51,0.047912,741116.93,0,2018,0.000000
4,47023,17097.0,343329.58,5158753.84,0.062400,343329.58,3,2015,0.000175
...,...,...,...,...,...,...,...,...,...
375,47155,95523.0,794897.48,27241866.58,0.028352,794897.48,38,2018,0.000398
376,47161,13248.0,61152.57,2100045.42,0.028296,61152.57,1,2015,0.000075
377,47161,13248.0,61152.57,2100045.42,0.028296,61152.57,1,2016,0.000075
378,47161,13248.0,61152.57,2100045.42,0.028296,61152.57,0,2017,0.000000


In [53]:
pop_result.sort_values(by='deaths_per_capita', ascending=False)

,fipscounty,population,money_spent_opioid_x,money_spent_non_opioid,opioid_ratio,money_spent_opioid_y,overdose_deaths,year,deaths_per_capita
315,47169,8773.0,217488.07,1382065.96,0.135968,217488.07,7,2018,0.000798
325,47027,7684.0,441764.67,4352417.61,0.092146,441764.67,6,2016,0.000781
266,47021,39713.0,421371.81,7098683.41,0.056033,421371.81,24,2017,0.000604
267,47021,39713.0,421371.81,7098683.41,0.056033,421371.81,21,2018,0.000529
341,47175,5675.0,40631.21,2020977.27,0.019708,40631.21,3,2016,0.000529
...,...,...,...,...,...,...,...,...,...
301,47137,5071.0,47070.64,7002481.97,0.006677,47070.64,0,2016,0.000000
46,47075,17944.0,117975.33,7143161.04,0.016248,117975.33,0,2017,0.000000
300,47137,5071.0,47070.64,7002481.97,0.006677,47070.64,0,2015,0.000000
93,47033,14554.0,80542.72,2793795.07,0.028021,80542.72,0,2016,0.000000


b. Which county has the most spending overall per capita?

In [63]:
pop_result['total_spending'] = pop_result['money_spent_opioid_x'] + pop_result['money_spent_non_opioid']
pop_result

,fipscounty,population,money_spent_opioid_x,money_spent_non_opioid,opioid_ratio,overdose_deaths,year,deaths_per_capita,total_spending
0,47017,28137.0,741116.93,14727057.51,0.047912,1,2015,0.000036,15468174.44
1,47017,28137.0,741116.93,14727057.51,0.047912,1,2016,0.000036,15468174.44
2,47017,28137.0,741116.93,14727057.51,0.047912,2,2017,0.000071,15468174.44
3,47017,28137.0,741116.93,14727057.51,0.047912,0,2018,0.000000,15468174.44
4,47023,17097.0,343329.58,5158753.84,0.062400,3,2015,0.000175,5502083.42
...,...,...,...,...,...,...,...,...,...
375,47155,95523.0,794897.48,27241866.58,0.028352,38,2018,0.000398,28036764.06
376,47161,13248.0,61152.57,2100045.42,0.028296,1,2015,0.000075,2161197.99
377,47161,13248.0,61152.57,2100045.42,0.028296,1,2016,0.000075,2161197.99
378,47161,13248.0,61152.57,2100045.42,0.028296,0,2017,0.000000,2161197.99


In [67]:
pop_result['total_spending_capita'] = pop_result['total_spending'] / pop_result['population']
pop_result.sort_values(by='total_spending_capita', ascending=False)

,fipscounty,population,money_spent_opioid_x,money_spent_non_opioid,opioid_ratio,overdose_deaths,year,deaths_per_capita,total_spending,total_spending_capita
301,47137,5071.0,47070.64,7.002482e+06,0.006677,0,2016,0.000000,7.049553e+06,1390.170106
303,47137,5071.0,47070.64,7.002482e+06,0.006677,1,2018,0.000197,7.049553e+06,1390.170106
300,47137,5071.0,47070.64,7.002482e+06,0.006677,0,2015,0.000000,7.049553e+06,1390.170106
302,47137,5071.0,47070.64,7.002482e+06,0.006677,1,2017,0.000197,7.049553e+06,1390.170106
186,47163,156519.0,5764963.90,1.710781e+08,0.032599,29,2017,0.000185,1.768431e+08,1129.850480
...,...,...,...,...,...,...,...,...,...,...
36,47121,11830.0,36781.95,1.344467e+06,0.026629,0,2015,0.000000,1.381249e+06,116.758167
100,47095,7588.0,5631.37,7.878898e+05,0.007097,0,2015,0.000000,7.935212e+05,104.575805
101,47095,7588.0,5631.37,7.878898e+05,0.007097,0,2016,0.000000,7.935212e+05,104.575805
102,47095,7588.0,5631.37,7.878898e+05,0.007097,0,2017,0.000000,7.935212e+05,104.575805


47137 spent the most with $1390 per capita

c. Which county has the most spending on opioids per capita?

--------------------------------------------------------------------------------------------------------------------------------------------

4. Unemployment

 a. Is there a correlation between unemployment rate and overdose deaths?

 b. Is there a correlation between unemployment and spending on opioids?

--------------------------------------------------------------------------------------------------------------------------------------------

5. Top prescribers

  a. Where are the top 10 opioid prescribers located?

  b. Who is the top prescriber in each county?

  c. What proportion of opioids are prescribed by the top 10 prescribers?  Top 50? Top 100?

--------------------------------------------------------------------------------------------------------------------------------------------

6. Nashville - Davidson County

a. Which zip codes in Davidson County have the most opioids prescribed?

  b. Any correlation between the number of missed trash pick ups and number of opioids prescribed?

--------------------------------------------------------------------------------------------------------------------------------------------